source/Preprocessor.cpp

```diff
@@ -395,28 +396,70 @@ Trivia Preprocessor::handleEndIfDirective(Token directive) {
     return parseBranchDirective(directive, Token(), taken);
 }
 
-void Preprocessor::expectTimescaleSpecifier(Token& value, Token& unit) {
-    // TODO: check for allowed values
-    auto token = peek();
-    if (token.kind == TokenKind::IntegerLiteral) {
+bool Preprocessor::expectTimescaleSpecifier(Token& value, Token& unit) {
+    if (peek().kind == TokenKind::IntegerLiteral) {
         value = consume();
         unit = expect(TokenKind::Identifier);
+    } else {
+        auto token = expect(TokenKind::TimeLiteral);
+        if (token.isMissing()) {
+            value = Token::createMissing(alloc, TokenKind::IntegerLiteral, token.location());
+            unit = Token::createMissing(alloc, TokenKind::Identifier, token.location());
+            return false;
+        }
+
+        const SVInt *val = std::get_if<SVInt>(&token.numericValue());
+        if (val == nullptr) {
+            addError(DiagCode::InvalidTimescaleSpecifier, token.location());
+            return false;
+        }
+        StringRef numText;
+        if      (*val == 1) numText = "1";
+        else if (*val == 10) numText = "10";
+        else if (*val == 100) numText = "100";
+        else {
+            addError(DiagCode::InvalidTimescaleSpecifier, token.location());
+            return false;
+        }
+
+        // Split the TimeLiteral into a IntegerLiteral and Identifer token
+        Token::Info* valueInfo = alloc.emplace<Token::Info>(token.trivia(),
+            numText, token.location(), token.getInfo()->flags | TokenFlags::IsFromPreprocessor);
```

Comment:
Hmm, we probably don't want to add the IsFromPreprocessor flag. The point of that is so that we can accurately reproduce the source that the user typed; anything from the preprocessor (such as from includes or macro expansion) wasn't actually there, so we hide it, but these tokens are actually there in the source.